File: app/src/main/java/com/chooloo/www/callmanager/ui/base/BaseBottomSheetDialogFragment.java
package com.chooloo.www.callmanager.ui.base;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.fragment.app.FragmentManager;

import com.chooloo.www.callmanager.databinding.FragmentBottomDialogBinding;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import static com.chooloo.www.callmanager.util.PermissionUtils.RC_DEFAULT;

public abstract class BaseBottomSheetDialogFragment extends BottomSheetDialogFragment implements MvpView {

    private boolean mIsShown = false;

    protected String[] mRequiredPermissions;

    protected BaseActivity mActivity;

    protected FragmentBottomDialogBinding binding;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof BaseActivity) {
            this.mActivity = (BaseActivity) context;
            mActivity.onAttachFragment(this);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentBottomDialogBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mRequiredPermissions = onGetPermissions();
        binding.bottomDialogFragmentCloseButton.setOnClickListener(view1 -> dismiss());
        onSetup();
    }

    @Override
    public void show(@NonNull FragmentManager manager, @Nullable String tag) {
        if (!mIsShown) {
            super.show(manager, tag);
            this.mIsShown = true;
        }
    }

    @Override
    public void dismiss() {
        if (mIsShown) {
            super.dismiss();
            this.mIsShown = false;
        }
    }

    @Override
    public String[] onGetPermissions() {
        return new String[]{};
    }

    @Override
    public boolean hasPermission(String permission) {
        return false;
    }

    @Override
    public boolean hasPermissions() {
        for (String permission : mRequiredPermissions) {
            if (!mActivity.hasPermission(permission)) {
                return false;
            }
        }
        return true;
    }

    @Override
    public void askForPermission(String permission, int requestCode) {
        askForPermissions(new String[]{permission}, requestCode);
    }

    @Override
    public void askForPermissions(String[] permissions, int requestCode) {
        requestPermissions(permissions, requestCode);
    }

    @Override
    public void askForPermissions() {
        askForPermissions(mRequiredPermissions, RC_DEFAULT);
    }

    @Override
    public void showMessage(String message) {
        if (mActivity != null) {
            mActivity.showMessage(message);
        }
    }

    @Override
    public void showMessage(@StringRes int stringResId) {
        if (mActivity != null) {
            mActivity.showMessage(stringResId);
        }
    }

    @Override
    public void showError(String message) {
        if (mActivity != null) {
            mActivity.showError(message);
        }
    }

    @Override
    public void showError(@StringRes int stringResId) {
        if (mActivity != null) {
            mActivity.showError(getString(stringResId));
        }
    }

    protected void putFragment(BaseFragment fragment) {
        getChildFragmentManager().beginTransaction().replace(binding.bottomDialogFragmentPlaceholder.getId(), fragment).commit();
    }

    protected Bundle getArgsSafely() {
        Bundle args = super.getArguments();
        if (args == null) {
            throw new IllegalArgumentException("You must create this fragment with newInstance()");
        }
        return args;
    }

    public boolean isShown() {
        return mIsShown;
    }
}

File: app/src/main/java/com/chooloo/www/koler/ui/base/BaseBottomSheetDialogFragment.kt
package com.chooloo.www.koler.ui.base

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.fragment.app.FragmentManager
import com.chooloo.www.koler.databinding.FragmentBottomDialogBinding
import com.google.android.material.bottomsheet.BottomSheetDialogFragment

abstract class BaseBottomSheetDialogFragment : BottomSheetDialogFragment(), MvpView {

    private lateinit var _binding: FragmentBottomDialogBinding
    protected lateinit var _activity: BaseActivity
    var isShown = false

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context !is BaseActivity) {
            throw TypeCastException("Fragment not a child of base activity")
        }
        _activity = context
        _activity.onAttachFragment(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentBottomDialogBinding.inflate(inflater, container, false)
        _binding.bottomDialogFragmentCloseButton.setOnClickListener { dismiss() }
        return _binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        onSetup()
    }

    override fun show(manager: FragmentManager, tag: String?) {
        if (!isShown) {
            super.show(manager, tag)
            isShown = true
        }
    }

    override fun dismiss() {
        if (isShown) {
            super.dismiss()
            isShown = false
        }
    }

    override fun hasPermission(permission: String): Boolean {
        return false
    }

    override fun hasPermissions(permissions: Array<String>): Boolean {
        return permissions.filter { p -> _activity.hasPermission(p) }.isNotEmpty()
    }

    override fun askForPermission(permission: String, requestCode: Int?) {
        askForPermissions(arrayOf(permission), requestCode)
    }

    override fun askForPermissions(permissions: Array<String>, requestCode: Int?) {
        requestPermissions(permissions, requestCode ?: 1)
    }

    override fun showMessage(message: String) {
        _activity.showMessage(message)
    }

    override fun showMessage(@StringRes stringResId: Int) {
        _activity.showMessage(stringResId)
    }

    override fun showError(message: String) {
        _activity.showError(message)
    }

    override fun showError(@StringRes stringResId: Int) {
        _activity.showError(getString(stringResId))
    }

    protected fun putFragment(fragment: BaseFragment) {
        childFragmentManager.beginTransaction().replace(_binding.bottomDialogFragmentPlaceholder.id, fragment).commit()
    }

    protected val argsSafely: Bundle
        get() = arguments
                ?: throw IllegalArgumentException("Always create fragment with newInstance()")

}